pkg/operator/upgradebackupcontroller/upgradebackupcontroller.go

```diff
@@ -0,0 +1,258 @@
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// UpgradeBackupController responds to an upgrade request to 4.9 by attempting
+// to ensure the cluster is backed up. The CVO is expected to wait on a ceo
+// condition indicating successful backup before responding to the upgrade
+// request. This is intended to ensure that an upgrade from 4.8 to 4.9 does not
+// proceed without a recent backup to ensure restore a cluster that is not
+// healthy after upgrade to 4.9.
+
+const (
+	backupConditionType       = "RecentBackup"
+	clusterBackupVersionLabel = "cluster-backup-version"
+)
+
+type UpgradeBackupController struct {
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	podLister            corev1listers.PodLister
+	nodeLister           corev1listers.NodeLister
+	networkLister        configv1listers.NetworkLister
+	clusterVersionLister configv1listers.ClusterVersionLister
+	targetImagePullSpec  string
+	operatorVersion      string
+}
+
+func NewUpgradeBackupController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	clusterVersionInformer configv1informers.ClusterVersionInformer,
+	networkInformer configv1informers.NetworkInformer,
+	eventRecorder events.Recorder,
+	targetImagePullSpec string,
+	operatorVersion string,
+) factory.Controller {
+	c := &UpgradeBackupController{
+		operatorClient:       operatorClient,
+		operatorVersion:      operatorVersion,
+		kubeClient:           kubeClient,
+		podLister:            kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
+		nodeLister:           kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+		networkLister:        networkInformer.Lister(),
+		clusterVersionLister: clusterVersionInformer.Lister(),
+		targetImagePullSpec:  targetImagePullSpec,
+	}
+	return factory.New().ResyncEvery(10*time.Minute).WithInformers(
+		kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Informer(),
+		kubeInformers.InformersFor("").Core().V1().Nodes().Informer(),
+		operatorClient.Informer(),
+		clusterVersionInformer.Informer(),
+	).WithSync(c.sync).ToController("ClusterBackupController", eventRecorder.WithComponentSuffix("cluster-backup-controller"))
+}
+
+func (c *UpgradeBackupController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	clusterVersion, err := c.clusterVersionLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	// Check cluster version status for backup condition
+	if !isRequireRecentBackup(clusterVersion) {
+		// Work is done
+		return nil
+	}
+
+	// Find the backup condition
+	_, status, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	backupVersion := c.operatorVersion
```

Comment:
How does the version come into this?  If we're currently taking a snapshot (as seen by an `Unknown` status on the current condition with an acceptably recent `lastTransitionTime` or some such), then we can keep waiting that backup out, right?  And if there is no backup running, or the running backup is too old, we start off a new one?  I'm not seeing a use-case for "5s ago I kicked off a snapshot, but now I see a different operator version, so kick off a new one".

Reply:
right version control is probably not necessary in this design as we are not really managing backups for all versions but instead only 4.9 -> 4.9.